### PR TITLE
Add trade station cart

### PR DESCRIPTION
**Pull request by liach**
More leftovers:
Trade stations can dupe items as a result of the recent inv refactor. Needs fix.
Trade stations guis have weird color if f3+b bounding box render is on
The AI is weird, dunno if it really works.
The initial bytes for the trade station cart cannot really be improved given the data parameters are initialized before the fields in classes are. Decoding from each default logic is expensive and no less ugly.

**The Issue**
See #1093 comment for trade station cart.
 
**The Proposal**
 - Added carts module for disabling carts more easily
 - Changed how minecarts open guis (simpler now)
 - Moved trade station logic to a shared logic class for easy cleanups except entity/te specific parts
 - 
 
**Possible Side Effects**
 - Affects cart management in the future in config (new module)
 - Affects trade station behavior
 
**Alternatives**
No alternative is present except leaving this feature. The trade station cart reduce the need of moving villagers, which move on their own and might disrupt the minecarts' movements.

### CODE REVIEW

**Comment by liach** [comment]
the compile error is caused by the missing dependencies for the test source set.

File: src/main/java/mods/railcraft/client/gui/FactoryGui.java [hunk 1]
```diff
@@ -105,7 +106,7 @@
                 case FEED_STATION:
                     return new GuiFeedStation(inv, (TileFeedStation) obj);
                 case TRADE_STATION:
-                    return new GuiTradeStation(inv, (TileTradeStation) obj);
+                    return new GuiTradeStation(inv, obj instanceof TileTradeStation ? ((TileTradeStation) obj).getLogic() : ((EntityCartTradeStation) obj).getLogic(), (IWorldNameable) obj);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/client/gui/FactoryGui.java]: Use an interface?

File: src/main/java/mods/railcraft/common/gui/containers/FactoryContainer.java [hunk 2]
```diff
@@ -92,7 +93,7 @@ public static Container build(EnumGui gui, InventoryPlayer inv, Object obj, Worl
                 case FEED_STATION:
                     return new ContainerFeedStation(inv, (TileFeedStation) obj);
                 case TRADE_STATION:
-                    return new ContainerTradeStation(inv, (TileTradeStation) obj);
+                    return new ContainerTradeStation(inv, obj instanceof TileTradeStation ? ((TileTradeStation) obj).getLogic() : ((EntityCartTradeStation) obj).getLogic());
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/gui/containers/FactoryContainer.java]: Again...interface

File: src/main/java/mods/railcraft/common/modules/ModuleExtras.java [hunk 3]
```diff
@@ -14,7 +14,7 @@
 import mods.railcraft.common.blocks.tracks.outfitted.TrackKits;
 import mods.railcraft.common.carts.RailcraftCarts;
 
-@RailcraftModule(value = "railcraft:extras", softDependencyClasses = ModuleTracks.class, description = "silly stuff, launch track, tnt carts, work cart")
+@RailcraftModule(value = "railcraft:extras", softDependencyClasses = {ModuleCarts.class, ModuleTracks.class}, description = "silly stuff, launch track, tnt carts, work cart")
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/modules/ModuleExtras.java]: Description needs an update.

File: src/main/java/mods/railcraft/common/util/entity/ai/EntityAIMoveToEntity.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import net.minecraft.entity.ai.RandomPositionGenerator;
+import net.minecraft.util.math.Vec3d;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Predicate;
+
+/**
+ * @author CovertJaguar <http://www.railcraft.info/>
+ */
+public class EntityAIMoveToEntity extends EntityAIBase {
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/util/entity/ai/EntityAIMoveToEntity.java]: Rename to EntityAISearchForEntity, and the block version to EntityAISearchForBlock.

**Review by CovertJaguar — CHANGES_REQUESTED** [review]
Looks good generally, though the datawatcher stuff is definitely iffy.

**Comment by liach** [comment]
wtf github removed that merge button here. stupid enough

**Review by CovertJaguar — APPROVED** [review]

**Comment by CovertJaguar** [comment]
FYI, your merge failed.

>/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/InventoryLogic.java:28: error: cannot find symbol
    public IInventory getInventory() {
           ^
  symbol:   class IInventory
  location: class InventoryLogic
/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/TradeStationLogic.java:58: error: incompatible types: IInventoryComposite cannot be converted to IInventory
    private final InventoryAdvanced recipeSlots = new InventoryAdvanced(9).callbackInv(inventory).phantom();
                                                                                       ^
/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/TradeStationLogic.java:64: error: incompatible types: IInventory cannot be converted to IInventoryComposite
        super(world, inventory);

**Comment by liach** [comment]
yeah, been lazy to fix it.

**Pull request merged**